platforms/esp/32/fastpin_esp32.h

```diff
@@ -24,68 +11,90 @@ template<uint8_t PIN, uint32_t MASK> class _ESPPIN {
+  inline static port_t hival() __attribute__ ((always_inline)) { return MASK; }
+  inline static port_t loval() __attribute__ ((always_inline)) { return MASK; }
+  inline static port_ptr_t port() __attribute__ ((always_inline)) { return 0; }
+
+  inline static port_ptr_t sport() __attribute__ ((always_inline)) { 
+      if (PIN < 32) return &GPIO.out_w1ts;
+      else if(PIN < 34) return &GPIO.out1_w1ts.val;
+      else return 0;
+  }
+
+  inline static port_ptr_t cport() __attribute__ ((always_inline)) {
+      if (PIN < 32) return &GPIO.out_w1tc;
+      else if(PIN < 34) return &GPIO.out1_w1tc.val;
+      else return 0;
+  }
+
   inline static port_t mask() __attribute__ ((always_inline)) { return MASK; }
 
   inline static bool isset() __attribute__ ((always_inline)) { return (0x004 & MASK); }
 };
 
-#define _DEFPIN_ESP32(PIN, REAL_PIN) template<> class FastPin<PIN> : public _ESPPIN<REAL_PIN, (1<<(REAL_PIN & 0xFF))> {};
-
+#define _DEFPIN_ESP32(PIN)  template<> class FastPin<PIN> : public _ESPPIN<PIN, ((uint32_t)1 << PIN)> {};
+#define _DEFPIN_32_33_ESP32(PIN) template<> class FastPin<PIN> : public _ESPPIN<PIN, ((uint32_t)1 << (PIN-32))> {};
 
-#ifdef FASTLED_ESP32_RAW_PIN_ORDER
+_DEFPIN_ESP32(0);
+// _DEFPIN_ESP32(1); Using TX causes flashiness when uploading
+_DEFPIN_ESP32(2); 
+// _DEFPIN_ESP32(3); Using RX causes flashiness when uploading
+_DEFPIN_ESP32(4);
+_DEFPIN_ESP32(5);
 
-_DEFPIN_ESP32(0,0); _DEFPIN_ESP32(1,1); _DEFPIN_ESP32(2,2); 
-_DEFPIN_ESP32(3,3); _DEFPIN_ESP32(4,4); _DEFPIN_ESP32(5,5); 
-
-// -- These are not safe to use:
+// -- These pins are not safe to use:
 // _DEFPIN_ESP32(6,6); _DEFPIN_ESP32(7,7); _DEFPIN_ESP32(8,8); 
 // _DEFPIN_ESP32(9,9); _DEFPIN_ESP32(10,10); _DEFPIN_ESP32(11,11); 
 
-_DEFPIN_ESP32(12,12); _DEFPIN_ESP32(13,13);
-_DEFPIN_ESP32(14,14); _DEFPIN_ESP32(15,15); _DEFPIN_ESP32(16,16);
-_DEFPIN_ESP32(17,17); _DEFPIN_ESP32(18,18); _DEFPIN_ESP32(19,19);
+_DEFPIN_ESP32(12);
+_DEFPIN_ESP32(13);
+_DEFPIN_ESP32(14);
+_DEFPIN_ESP32(15);
+_DEFPIN_ESP32(16);
+_DEFPIN_ESP32(17);
+_DEFPIN_ESP32(18);
+_DEFPIN_ESP32(19);
 
 // No pin 20 : _DEFPIN_ESP32(20,20); 
 
-_DEFPIN_ESP32(21,21); _DEFPIN_ESP32(22,22); _DEFPIN_ESP32(23,23); 
+_DEFPIN_ESP32(21); // Works, but note that GPIO21 is I2C SDA
+_DEFPIN_ESP32(22); // Works, but note that GPIO22 is I2C SCL
+_DEFPIN_ESP32(23); 
 
 // No pin 24 : _DEFPIN_ESP32(24,24); 
 
-_DEFPIN_ESP32(25,25); _DEFPIN_ESP32(26,26); _DEFPIN_ESP32(27,27); 
+_DEFPIN_ESP32(25);
+_DEFPIN_ESP32(26);
+_DEFPIN_ESP32(27); 
 
 // No pin 28-31: _DEFPIN_ESP32(28,28); _DEFPIN_ESP32(29,29); _DEFPIN_ESP32(30,30); _DEFPIN_ESP32(31,31);
 
 // Need special handling for pins > 31
-// _DEFPIN_ESP32(32,32); _DEFPIN_ESP32(33,33);
-
-#define PORTA_FIRST_PIN 32
-// The rest of the pins - these are generally not available
-// _DEFPIN_ESP32(11,6);
-// _DEFPIN_ESP32(12,7); _DEFPIN_ESP32(13,8); _DEFPIN_ESP32(14,9); _DEFPIN_ESP32(15,10);
-// _DEFPIN_ESP32(16,11);
-
-#endif
+_DEFPIN_32_33_ESP32(32); 
```

Comment:
Note that according to https://github.com/espressif/esp-idf/blob/master/components/soc/esp32/include/soc/gpio_struct.h - this can go up to pin 39 - even if no boards are exposing anything above 32/33 yet.

Reply:
Pins 34-39 are input only.

Reply:
For now : )